File: 02_double_dispatch/erlang/paper_scissor_rock.erl
#!/usr/bin/env escript

% We’ll use  atoms  rock  paper scissor
% to represent the three choices.
beats(paper, scissor) -> "paper loses";
beats(paper, rock) -> "paper wins";
beats(scissor, paper) -> "scissor wins";
beats(scissor, rock) -> "scissor loses";
beats(rock, paper) -> "rock wins";
beats(rock, scissor) -> "rock loses";
beats(E, E) -> "draw".
  
main([]) ->
  io:format("~s ~s~n", ["paper beats paper =", beats(paper, paper)]),
  io:format("~s ~s~n", ["paper beats scissor =", beats(paper, scissor)]),
  io:format("~s ~s~n", ["paper beats rock =", beats(paper, rock)]),
  io:format("~s ~s~n", ["scissor beats paper =", beats(scissor, paper)]),
  io:format("~s ~s~n", ["scissor beats scissor =", beats(scissor, scissor)]),
  io:format("~s ~s~n", ["scissor beats rock =", beats(scissor, rock)]),
  io:format("~s ~s~n", ["rock beats paper =", beats(rock, paper)]),
  io:format("~s ~s~n", ["rock beats scissor =", beats(rock, scissor)]),
  io:format("~s ~s~n", ["rock beats rock =", beats(rock, rock)]).
  

File: 02_double_dispatch/clojure/paper_scissor_rock.clj
; Multi-methods can be used to solve this problem
; Here is our dispatch function
(defmulti beats
  (fn [e1 e2] [e1 e2]))

; We’ll use keys :rock  :paper and :scissor
; to represent the three choices.
(defmethod beats [:paper :scissor] [e1 e2] "paper loses")
(defmethod beats [:paper :rock] [e1 e2] "paper wins")
(defmethod beats [:scissor :paper] [e1 e2] "scissor wins")
(defmethod beats [:scissor :rock] [e1 e2] "scissor loses")
(defmethod beats [:rock :paper] [e1 e2] "rock loses")
(defmethod beats [:rock :scissor] [e1 e2] "rock wins")
(defmethod beats :default [e1 e2] "draw")

(println "paper beats paper = "     (beats :paper :paper))
(println "paper beats scissor = "   (beats :paper :scissor))
(println "paper beats rock = "      (beats :paper :rock))
(println "scissor beats paper = "   (beats :scissor :paper))
(println "scissor beats scissor = " (beats :scissor :scissor))
(println "scissor beats rock = "    (beats :scissor :rock))
(println "rock beats paper = "      (beats :rock :paper))
(println "rock beats scissor = "    (beats :rock :scissor))
(println "rock beats rock = "       (beats :rock :rock))